data/trie/patriciaMerkleTrie.go

```diff
@@ -451,11 +469,18 @@ func (tr *patriciaMerkleTrie) Snapshot() error {
 	tr.mutOperation.Lock()
 	defer tr.mutOperation.Unlock()
 
-	err := tr.root.commit(false, 0, tr.db, tr.marshalizer, tr.hasher)
+	if tr.snapshotInProgress {
```

Comment:
mutex protection should be enough. make sure that it stops there, if it is called the 2nd time.

Reply:
After the `Snapshot()` func is done, the mutex is released, but the `snapshot(newTrie *patriciaMerkleTrie, db data.DBWriteCacher)` is still in progress and is not protected by a mutex.